doctest's spec tests failing with GHC 9

There are some test failures when building with GHC 9. Eg:

```
# stack8.10.yaml
resolver: lts-18.0
packages:
- .
```

```
# stack.yaml
resolver: nightly-2021-07-16
packages:
- .
```

With ghc 8:
```
$ stack test doctest:test:spec --stack-yaml stack8.10.yaml
...
Finished in 82.0428 seconds
170 examples, 0 failures

doctest> Test suite spec passed
```
With ghc 9:
```
$ stack test doctest:test:spec
...
Failures:

  test/MainSpec.hs:94:7: 
  1) Main.doctest template-haskell
       template-haskell ["Foo.hs"]
       expected: Examples: 2  Tried: 2  Errors: 0  Failures: 0
        but got: Examples: 2  Tried: 2  Errors: 0  Failures: 1

  To rerun use: --match "/Main/doctest/template-haskell/"

  test/MainSpec.hs:148:7: 
  2) Main, doctest as a runner for QuickCheck properties, runs $setup before each property
       property-setup ["Foo.hs"]
       expected: Examples: 3  Tried: 3  Errors: 0  Failures: 0
        but got: Examples: 3  Tried: 3  Errors: 0  Failures: 1

  To rerun use: --match "/Main/doctest as a runner for QuickCheck properties/runs $setup before each property/"

  test/MainSpec.hs:177:7: 
  3) Main, doctest (regression tests), template-haskell-bugfix
       template-haskell-bugfix ["Main.hs"]
       expected: Examples: 2  Tried: 2  Errors: 0  Failures: 0
        but got: Examples: 2  Tried: 2  Errors: 0  Failures: 1

  To rerun use: --match "/Main/doctest (regression tests)/template-haskell-bugfix/"

  test/RunSpec.hs:108:5: 
  4) Run.doctest can deal with potentially problematic GHC options
       uncaught exception: ExitCode
       ExitFailure 1

  To rerun use: --match "/Run/doctest/can deal with potentially problematic GHC options/"

  test/RunSpec.hs:128:11: 
  5) Run.doctestWithOptions, on parse error, prints a useful error message
       expected: "\nFoo.hs:6:1: error:\n    parse error (possibly incorrect indentation or mismatched brackets)\n"
        but got: "\n\ESC[;1mFoo.hs:6:1: \ESC[;1m\ESC[31merror:\ESC[0m\ESC[0m\ESC[;1m\ESC[0m\ESC[0m\ESC[;1m\n    parse error (possibly incorrect indentation or mismatched brackets)\ESC[0m\ESC[0m\n"

  To rerun use: --match "/Run/doctestWithOptions/on parse error/prints a useful error message/"

Randomized with seed 2096325755

Finished in 90.4501 seconds
170 examples, 5 failures
```

--- a/README.markdown
+++ b/README.markdown
@@ -352,6 +352,32 @@ Haddock](#hiding-examples-from-haddock), e.g.:
 
 [language-pragma]: http://www.haskell.org/ghc/docs/latest/html/users_guide/pragmas.html#language-pragma
 
+### Limitations
+
+Due to [a GHC bug](https://gitlab.haskell.org/ghc/ghc/-/issues/20670), running
+`:set -XTemplateHaskell` within `ghci` may unload any modules that were
+specified on the command line.
+
+To address this `doctest >= 0.19.0` does two things:
+
+1. Doctest always enables `-XTemplateHaskell`.  So it is safe to use Template
+   Haskell in examples without enabling the extension explicitly.
+1. Doctest filters out `-XTemplateHaskell` from single-line `:set`-statements.
+   So it is still safe to include `:set -XTemplateHaskell` in examples for
+   documentation purposes.  It may just not work as intended in `ghci` due to
+   that GHC bug.
+
+Doctest does not filter out `-XTemplateHaskell` from multi-line
+`:set`-statements.  So if you e.g. use
+
+```
+>>> :{
+:set -XTemplateHaskell
+:}
+```
+then you are on your own.
+
+
 ### Cabal integration
 
 Doctest provides both, an executable and a library.  The library exposes a

--- a/doctest.cabal
+++ b/doctest.cabal
@@ -1,10 +1,8 @@
-cabal-version: 1.12
+cabal-version: 2.0
 
 -- This file has been generated from package.yaml by hpack version 0.34.4.
 --
 -- see: https://github.com/sol/hpack
---
--- hash: d72341386fa5019d836d102ce53149841ad11aa7bd786c3ba9b8cc28ca3c9046
 
 name:           doctest
 version:        0.18.2

--- a/src/Interpreter.hs
+++ b/src/Interpreter.hs
@@ -11,6 +11,7 @@ module Interpreter (
 -- exported for testing
 , ghcInfo
 , haveInterpreterKey
+, filterExpression
 ) where
 
 import           System.Process
@@ -55,12 +56,16 @@ withInterpreter flags action = do
   let
     args = flags ++ [
         "--interactive"
+      , xTemplateHaskell
 #if __GLASGOW_HASKELL__ >= 802
       , "-fdiagnostics-color=never"
       , "-fno-diagnostics-show-caret"
 #endif
       ]
   bracket (new defaultConfig{configGhci = ghc} args) close action
+
+xTemplateHaskell :: String
+xTemplateHaskell = "-XTemplateHaskell"
 
 -- | Evaluate an expression; return a Left value on exceptions.
 --
@@ -75,11 +80,17 @@ filterExpression :: String -> Either String String
 filterExpression e =
   case lines e of
     [] -> Right e
-    l  -> if firstLine == ":{" && lastLine /= ":}" then fail_ else Right e
+    l  -> if firstLine == ":{" && lastLine /= ":}" then err else Right (filterXTemplateHaskell e)
       where
         firstLine = strip $ head l
         lastLine  = strip $ last l
-        fail_ = Left "unterminated multiline command"
+        err = Left "unterminated multi-line command"
   where
     strip :: String -> String
     strip = dropWhile isSpace . reverse . dropWhile isSpace . reverse
+
+filterXTemplateHaskell :: String -> String
+filterXTemplateHaskell input = case words input of
+  [":set", setting] | setting == xTemplateHaskell -> ""
+  ":set" : xs | xTemplateHaskell `elem` xs -> unwords $ ":set" : filter (/= xTemplateHaskell) xs
+  _ -> input

--- a/src/Property.hs
+++ b/src/Property.hs
@@ -30,7 +30,6 @@ runProperty repl expression = do
   _ <- Interpreter.safeEval repl "import Test.QuickCheck ((==>))"
   _ <- Interpreter.safeEval repl "import Test.QuickCheck.All (polyQuickCheck)"
   _ <- Interpreter.safeEval repl "import Language.Haskell.TH (mkName)"
-  _ <- Interpreter.safeEval repl ":set -XTemplateHaskell"
   r <- freeVariables repl expression >>=
        (Interpreter.safeEval repl . quickCheck expression)
   case r of

--- a/test/InterpreterSpec.hs
+++ b/test/InterpreterSpec.hs
@@ -5,7 +5,7 @@ import           Prelude.Compat
 
 import           Test.Hspec
 
-import           Interpreter (interpreterSupported, haveInterpreterKey, ghcInfo, withInterpreter, safeEval)
+import           Interpreter (interpreterSupported, haveInterpreterKey, ghcInfo, withInterpreter, safeEval, filterExpression)
 
 main :: IO ()
 main = hspec spec
@@ -27,4 +27,14 @@ spec = do
       Interpreter.safeEval ghci "23 + 42" `shouldReturn` Right "65\n"
 
     it "returns Left on unterminated multiline command" $ withInterpreter [] $ \ghci -> do
-      Interpreter.safeEval ghci ":{\n23 + 42" `shouldReturn` Left "unterminated multiline command"
+      Interpreter.safeEval ghci ":{\n23 + 42" `shouldReturn` Left "unterminated multi-line command"
+
+  describe "filterExpression" $ do
+    it "removes :set -XTemplateHaskell" $ do
+      filterExpression ":set -XTemplateHaskell" `shouldBe` Right ""
+
+    it "filters -XTemplateHaskell" $ do
+      filterExpression ":set -XTemplateHaskell -XCPP" `shouldBe` Right ":set -XCPP"
+
+    it "leaves :set-statement that do not set -XTemplateHaskell alone " $ do
+      filterExpression ":set -XFoo   -XBar" `shouldBe` Right ":set -XFoo   -XBar"

